servers/physics_2d/body_2d_sw.cpp

```diff
@@ -52,9 +57,27 @@ void Body2DSW::update_inertias() {
 			//update tensor for allshapes, not the best way but should be somehow OK. (inspired from bullet)
 			real_t total_area = 0;
 
+			for (int i = 0; i < get_shape_count(); i++) {
+				total_area += get_shape_area(i);
+			}
+
+			center_of_mass_local.x = center_of_mass_local.y = 0;
+
 			for (int i = 0; i < get_shape_count(); i++) {
 
-				total_area += get_shape_aabb(i).get_area();
+				if (is_shape_disabled(i)) {
+					continue;
+				}
+				real_t area = get_shape_area(i);
+				real_t mass = area * this->mass / total_area;
+				Shape2DSW *shape = get_shape(i);
+
+				if (dynamic_cast<ConvexPolygonShape2DSW *>(shape) || dynamic_cast<ConcavePolygonShape2DSW *>(shape)) {
```

Comment:
Godot uses its own `Object::cast_to<ConvexPolygonShape2DSW *>` for dynamic cast.

Reply:
Because `Shape2DSW` don't inherit class `Object` so I can't use it.

Reply:
Still, we can't use `dynamic_cast` as some platforms are built without RTTI (Android, HTML5).
Not sure how it should be done in practice, I see there's `SAFE_CAST` but it's only used in `editor/collada/collada.cpp` so I don't know if it's the recommended approach either.

Reply:
Maybe adding a function `is_center_zero()` like this?
https://github.com/godotengine/godot/blob/ab9c14cca63757ff47dfc6c38b35baf2a32a51ac/servers/physics_2d/shape_2d_sw.h#L80

Reply:
Maybe, but perhaps `get_center_offset` could perhaps simplify calculations down the road:
```cpp
// shape_2d
// could also compute polygons center of mass here...
virtual Vector2 get_center_offset() const { return Vector2(); }

// collision_object_2d
_FORCE_INLINE_ const Transform2D &get_shape_center_offset(int p_index) const {
	CRASH_BAD_INDEX(p_index, shapes.size());
	return shapes[p_index].center_offset;
}
// body_2d
center_of_mass_local += mass * get_shape_center_offset(i);
```